Add scanner to scan dirs and files of a user and save the info into database.

fkztw opened the pull request:
Also made some modifications to files/models.py for a better table schema.

Commit:
Add File node for files/models.py

Commit:
Remove @property name of Directory for files/models.py.

Name of the Directory should be set when the scanner stores the
Directory into table.

Commit:
Add migrations for File node of files/models.py

Commit:
Use auto_now_add for created and auto_now for updated for Inode.

Commit:
Add on_delete CASCADE to UserProfile for User.

and change get_home_dir() to @property home_dir_path.

Commit:
Add scanner.

Scan the USER_ROOT_DIR of a specific user, and save directories and
files info into database.

Use `$ python manage.py scan $username` to execute.

pellaeon commented on files/models.py:
```diff
@@ -26,8 +24,8 @@ class Inode(FYNode):
     """
     owner = models.ForeignKey(User)
-    created = models.DateTimeField(default=timezone.now)
-    updated = models.DateTimeField(default=timezone.now)
```
The reason I did not use `auto_now_add` is that, as we have discussed, model fields should serve as a cache for underlying file object's fields, so their data should be the same. If we use `auto_now_add` that would mean `created` will be set to when the model object was created, not the actual `ctime` of the file.
As for `updated`, setting `auto_now` will make the value of the field always set to the time you call `.save()`, but in some scenario we might simply want to update its metadata, such as MIME type, in this case `updated` should still be the file's `mtime` instead of the time when metadata is saved.

ps. perhaps `updated` should be renamed to `modified` to indicate its relationship with `mtime`

fkztw replied:
oh, I see.
so

``` python
 created = models.DateTimeField(default=timezone.now)
modified = models.DateTimeField(default=timezone.now)
```

And we have to update the value of `modified` whenever the `mtime` of the file object changes?

pellaeon replied:
Yes.

For now, there's no way `mtime` could possibly be changed after the initial file upload, since:
1. HTTP file upload won't send original file metadata, so we can only set `mtime` to the time it is uploaded
2. there's no Feng Yuan app to change the file content

But `mtime` will be needed in the future. It will be updated in these cases:
1. a client capable of sending original file metadata
2. other server side programs (such as FTP server) may change the file content

pellaeon commented:
And perhaps some tests on the scanner would be nice?

fkztw commented:
https://docs.djangoproject.com/en/1.10/topics/testing/tools/#topics-testing-management-commands

pellaeon commented:
You're familiar with TravisCI ? You can set it up along the way, that would be faster, I have no experience setting up TravisCI.

Commit:
Rename USER_ROOT_DIR_NAME to USER_FILES_DIR_NAME.

Get the path of user_dir from user.profile.home_dir_path.

Commit:
Add signals to User and UserProfile.

+ Create UserProfile after creating User.
+ Create user_dir after creating UserProfile.
+ Delete user_dir before deleting UserProfile.

Commit:
[#1] Add setUp and tearDown method for testing of management command 'scan'

Commit:
[#1] Add test for file and dir should not be scanned for scanner.

Commit:
[#1] Add test for files and dirs should be scanned for scanner.

Commit:
Modify files.models for coding style.

Commit:
[#1] Make scanner record the parent of Directory.

Commit:
[#1] Add testings for management command 'scan' (scanner)

+ test files and dirs should be scanned
+ test files and dirs should not be scanned
+ test owner of files and dirs
+ test updated_time
+ test created_time
+ test parent of node

fkztw commented:
Still need a `.travis.yml`

Commit:
Add a simple .travis.yml

Commit:
Add more Python versions into .travis.yml.

Commit:
Add pytest as testing framework.

Use with pytest-django and pytest-cov.

fkztw commented:
Looks good.

Commit:
Rename files/tests/test_scanner.py back to files/tests.py.

To obey Django default settings for djano-admin.py startapp

pellaeon commented:
Nice, you can merge if there's no more commits to add.

Commit:
Revert "Use auto_now_add for created and auto_now for updated for Inode."

This reverts commit d5e62193c2236fcd0a7ec300d5ffe209124bda34.

django timezone.now will save timezone-aware datetime object into db
when `USE_TZ = True` (which is default) in the settings.py

Besides, use `auto_now` and `auto_now_add` will make this field cannot
be overrided in the future.

We have to save the `created` and `updated` with the ctime and mtime of
the inode object in the file system in the future, so we should use
`default` not `auto_now` and `auto_now_add`.

ref:
https://docs.djangoproject.com/en/1.10/ref/models/fields/#django.db.models.DateField.auto_now_add

Commit:
Add migrations for commit e9383bd.
